src/test/scala/com/microsoft/hyperspace/index/E2EHyperspaceRulesTests.scala

```diff
@@ -532,14 +527,80 @@ class E2EHyperspaceRulesTests extends HyperspaceSuite with SQLHelper {
       assert(planRootPaths.equals(Seq(PathUtils.makeAbsolute(location))))
 
       // Refresh the index to remove deleted source data file records from index.
-      hyperspace.refreshIndex(indexConfig.indexName)
+      hyperspace.refreshIndex(indexConfig.indexName, REFRESH_MODE_INCREMENTAL)
 
       // Verify index usage on latest version of index (v=1) after refresh.
       verifyIndexUsage(query2, getIndexFilesPath(indexConfig.indexName, Seq(1)))
       FileUtils.delete(dataPath)
     }
   }
 
+  test(
+    "Verify JoinIndexRule utilizes indexes correctly after incremental refresh when some file " +
+      "gets deleted and some appended to source data.") {
+    withTempPathAsString { testPath =>
+      withSQLConf(IndexConstants.INDEX_LINEAGE_ENABLED -> "true") {
+        // Setup. Create data.
+        val indexConfig = IndexConfig("index", Seq("c2"), Seq("c4"))
+        import spark.implicits._
+        SampleData.testData
+          .toDF("c1", "c2", "c3", "c4", "c5")
+          .limit(10)
+          .write
+          .parquet(testPath)
+        val df = spark.read.load(testPath)
+
+        // Create index.
+        hyperspace.createIndex(df, indexConfig)
+
+        // Delete some source data file.
+        val dataPath = new Path(testPath, "*parquet")
+        val dataFileNames = dataPath
+          .getFileSystem(new Configuration)
+          .globStatus(dataPath)
+          .map(_.getPath)
+
+        assert(dataFileNames.nonEmpty)
+        val fileToDelete = dataFileNames.head
+        FileUtils.delete(fileToDelete)
```

Comment:
Can you create a helper function and use it across tests (I see this logic is duplicated)? Something like:
```scala
// Returns the paths of deleted files.
def deleteFiles(path: String, pattern: String, numFilesToDelete: Int): Seq[Path]
```

Reply:
thanks @imback82 , added